pkg/ingester/ingester_forced_compaction_test.go

```diff
@@ -0,0 +1,600 @@
+)
+
+func TestIngester_compactBlocksToReduceInMemorySeries_ShouldCompactHeadUpUntilNowMinusActiveSeriesMetricsIdleTimeout(t *testing.T) {
+	var (
+		ctx          = context.Background()
+		ctxWithUser  = user.InjectOrgID(ctx, userID)
+		metricName   = "metric_1"
+		metricLabels = labels.FromStrings(labels.MetricName, metricName)
+		metricModel  = map[model.LabelName]model.LabelValue{model.MetricNameLabel: model.LabelValue(metricName)}
+		now          = time.Now()
+		sampleTimes  []time.Time
+	)
+
+	cfg := defaultIngesterTestConfig(t)
+	cfg.ActiveSeriesMetrics.Enabled = true
+	cfg.ActiveSeriesMetrics.IdleTimeout = 20 * time.Minute
+	cfg.BlocksStorageConfig.TSDB.HeadCompactionInterval = time.Hour // Do not trigger it during the test, so that we trigger it manually.
+	cfg.BlocksStorageConfig.TSDB.ForcedHeadCompactionMinInMemorySeries = 1
+	cfg.BlocksStorageConfig.TSDB.ForcedHeadCompactionMinEstimatedSeriesReductionPercentage = 0
+
+	ingester, err := prepareIngesterWithBlocksStorage(t, cfg, nil)
+	require.NoError(t, err)
+	require.NoError(t, services.StartAndAwaitRunning(ctx, ingester))
+	t.Cleanup(func() {
+		require.NoError(t, services.StopAndAwaitTerminated(ctx, ingester))
+	})
+
+	// Wait until it's ACTIVE.
+	test.Poll(t, time.Second, ring.ACTIVE, func() interface{} {
+		return ingester.lifecycler.GetState()
+	})
+
+	userBlocksDir := filepath.Join(ingester.cfg.BlocksStorageConfig.TSDB.Dir, userID)
+
+	t.Run("push a series and force TSDB head compaction", func(t *testing.T) {
+		// Push a series with a sample.
+		sampleTime := now
+		sampleTimes = append(sampleTimes, sampleTime)
+		require.NoError(t, pushSeriesToIngester(ctxWithUser, t, ingester, []series{{metricLabels, 1.0, sampleTime.UnixMilli()}}))
+
+		// Advance time and then check if TSDB head compaction should be forced.
+		// We expect no block to be created because there's no sample before "now - active series idle timeout".
+		now = now.Add(10 * time.Minute)
+		ingester.compactBlocksToReduceInMemorySeries(ctx, now)
+		require.Len(t, listBlocksInDir(t, userBlocksDir), 0)
+
+		// Further advance time and then check if TSDB head compaction should be forced.
+		// The previously written sample is expected to be compacted.
+		now = now.Add(11 * time.Minute)
+		ingester.compactBlocksToReduceInMemorySeries(ctx, now)
+
+		require.Len(t, listBlocksInDir(t, userBlocksDir), 1)
+		newBlockID := listBlocksInDir(t, userBlocksDir)[0]
+
+		assert.Equal(t, model.Matrix{
+			{
+				Metric: metricModel,
+				Values: []model.SamplePair{{Timestamp: model.Time(sampleTime.UnixMilli()), Value: 1.0}},
+			},
+		}, readMetricSamplesFromBlockDir(t, filepath.Join(userBlocksDir, newBlockID.String()), metricName))
+
+		// We expect the series to be dropped from TSDB Head because there's was no sample more recent than
+		// "now - active series idle timeout".
+		db := ingester.getTSDB(userID)
+		require.NotNil(t, db)
+		assert.Equal(t, uint64(0), db.Head().NumSeries())
```

Comment:
I don't think it will tell us much, because active series are always 0. The problem is that time is currently not mockable for the `Push()` request, so active series usage is always tracked with `time.Now()` and not the time mocked in the test.

Reply:
👍